### PR TITLE
Add Docs.undocumented_names

### LINKED ISSUE #51174: add a way to test whether exported symbols are documented
As [discussed on discourse](https://discourse.julialang.org/t/check-if-a-function-has-a-docstring/103489), it would be nice to have a (documented) way to test whether all exported symbols in a module have docstrings.  Currently, this seems to require undocumented internals?

Something like:
```jl
# hasdoc should go in Base.Docs; might also be accomplished by refactoring Base.Docs.doc somehow
using  Base.Docs: Binding, defined, getdoc, resolve, meta, modules, aliasof
hasdoc(mod::Module, sym::Symbol) = hasdoc(Base.Docs.Binding(mod, sym))
function hasdoc(binding::Binding, sig::Type = Union{})
    defined(binding) && !isnothing(getdoc(resolve(binding), sig)) && return true
    for mod in modules
        dict = meta(mod; autoinit=false)
        !isnothing(dict) && haskey(dict, binding) && return true
    end
    alias = aliasof(binding)
    return alias == binding ? false : hasdoc(alias, sig)
end

# possible function for the Test module
function check_documented(mod::Module)
    nodocs = filter!(sym -> !hasdoc(mod, sym), names(mod))
    isempty(nodocs) || error("missing docstrings in $mod: $nodocs")
end
```
For example, this catches some missing docstrings in `Base` and several standard libraries:
```jl
julia> check_documented(Base)
ERROR: missing docstrings in Base: [:CanonicalIndexError, :CapturedException, :InvalidStateException]

julia> check_documented(LinearAlgebra)
ERROR: missing docstrings in LinearAlgebra: [:ColumnNorm, :LAPACKException, :NoPivot, :RankDeficientException, :RowMaximum, :RowNonZero, :copy_transpose!]

julia> check_documented(Markdown)
ERROR: missing docstrings in Markdown: [Symbol("@doc_str"), Symbol("@md_str"), :html, :latex]

julia> check_documented(Pkg)
ERROR: missing docstrings in Pkg: [Symbol("@pkg_str"), :PKGMODE_MANIFEST, :PKGMODE_PROJECT, :PRESERVE_ALL, :PRESERVE_ALL_INSTALLED, :PRESERVE_DIRECT, :PRESERVE_NONE, :PRESERVE_SEMVER, :PRESERVE_TIERED, :PRESERVE_TIERED_INSTALLED, :Pkg, :PreserveLevel, :Registry, :UPLEVEL_MAJOR, :UPLEVEL_MINOR, :UPLEVEL_PATCH]
```

### PR DESCRIPTION
**Pull request by jariji**
Fixes #51174

Cross-ref: #52139

### CODE REVIEW

**Comment by jariji** [comment]
@stevengj ready for review.

File: base/docs/Docs.jl [hunk 1]
```diff
@@ -675,4 +675,15 @@ function hasdoc(binding::Docs.Binding, sig::Type = Union{})
 end
 
 
+"""
+    check_documented(mod::Module; all=false)
+
+Check all names in the module are documented. `all` determines which names are checked,
+following the behavior of `names(mod; all)`.
```
Comment (stevengj) [inline base/docs/Docs.jl:682]: Should have a cross-reference to names docs. Should probably just repeat what all means here

**Comment by stevengj** [comment]
I suspect we don’t want a function that throws an error here. We probably just want a function that returns a sorted list of undocumented symbols. Then people can just add tests like
```
@test Docs.undocumented_names(MyMod) == [:known1, :known2]
```

File: base/docs/Docs.jl [hunk 2]
```diff
@@ -676,14 +676,15 @@ end
 
 
 """
-    check_documented(mod::Module; all=false)
+    undocumented_names(mod::Module; all=false)
 
-Check all names in the module are documented. `all` determines which names are checked,
-following the behavior of `names(mod; all)`.
+Return the undocumented names in `module`. `all=false` returns only public names;
+`all=true` also includes nonpublic names, following the behavior of [`names`](@ref).
```
Comment (stevengj) [inline base/docs/Docs.jl:682]:
```suggestion
Return an array of undocumented symbols in `module` (that is, lacking docstrings).
`all=false` returns only exported symbols; whereas `all=true` also includes
non-exported symbols, following the behavior of [`names`](@ref).
```

File: base/docs/Docs.jl [hunk 3]
```diff
@@ -676,14 +676,15 @@ end
...
+
+See also: [`names`](@ref), [`Docs.hasdoc`](@ref).
 """
-function check_documented(mod::Module; all=false)
-    nodocs = filter!(sym -> !hasdoc(mod, sym), names(mod; all))
-    isempty(nodocs) || error("missing docstrings in $mod: $nodocs")
+function undocumented_names(mod::Module; all=false)
+    filter(!hasdoc, names(mod; all))
 end
```
Comment (stevengj) [inline base/docs/Docs.jl]:
```suggestion
undocumented_names(mod::Module; all=false) = filter(!hasdoc, names(mod; all))
```

Comment (jariji) [reply base/docs/Docs.jl]: That doesn't work because `names` returns `Symbols` not `Binding`s.

**Comment by jariji** [comment]
```
Expression: Docs.undocumented_names(_ModuleWithSomeDocumentedNames; all = true) == [:g]
   Evaluated: [Symbol("##meta#47"), Symbol("#eval"), Symbol("#f"), Symbol("#g"), Symbol("#include"), :eval, :g, :include] == [:g]
```

What should happen here? @stevengj

**Comment by stevengj** [comment]
`Symbol("##meta#47"), Symbol("#eval"), Symbol("#f"), Symbol("#g"), Symbol("#include")` are internal names generated by Julia's compiler.

For now, if you want to check the `all=true` case, I would filter the symbols for `!Base.isidentifier` to avoid involving compiler internals in the test.   In the future, we may want to add an `isidentifier::Bool=false` option to `names` as well as this function in order to filter out non-identifier symbols by default.

File: base/docs/Docs.jl [hunk 4]
```diff
@@ -675,4 +675,19 @@ function hasdoc(binding::Docs.Binding, sig::Type = Union{})
...
+Return an array of undocumented symbols in `module` (that is, lacking docstrings).
+`all=false` returns only exported symbols; whereas `all=true` also includes
+non-exported symbols, following the behavior of [`names`](@ref).
+
+See also: [`names`](@ref), [`Docs.hasdoc`](@ref).
+"""
+function undocumented_names(mod::Module; all=false)
+    filter(names(mod; all)) do sym
+        !hasdoc(mod, sym)
+    end
```
Comment (stevengj) [inline base/docs/Docs.jl:690]:
```suggestion
function undocumented_names(mod::Module; all::Bool=false, identifiers::Bool=true)
    filter!(names(mod; all)) do sym
        !hasdoc(mod, sym) && (!identifiers || Base.isidentifier(sym))
    end
```
Maybe add an option here to filter out symbols that are not valid identifiers by default.

File: doc/src/manual/documentation.md [hunk 5]
```diff
@@ -20,7 +20,8 @@ environments provide a way to access documentation directly:
   under the cursor.
 
 
-`Docs.hasdoc(module, name)::Bool` tells whether a name has a docstring.
+`Docs.hasdoc(module, name)::Bool` tells whether a name has a docstring. `Docs.undocumented_names(module; all)`
+returns the undocumented names in a module.
```
Comment (stevengj) [inline doc/src/manual/documentation.md:24]: These should also be added to the index of the reference manual somewhere.

File: test/docs.jl [hunk 6]
```diff
@@ -76,6 +76,25 @@ function break_me_docs end
...
+module _ModuleWithSomeDocumentedNames
+export f
+"f() is 1."
+f() = 1
+g() = 2
+end
+
+@test Docs.undocumented_names(_ModuleWithUndocumentedNames) == [:f]
+@test isempty(Docs.undocumented_names(_ModuleWithSomeDocumentedNames))
+@test Docs.undocumented_names(_ModuleWithSomeDocumentedNames; all=true) == [:g]
```
Comment (stevengj) [inline test/docs.jl:95]:
```suggestion
@test Docs.undocumented_names(_ModuleWithSomeDocumentedNames; all=true) == [:eval, :g]
```
with the above change, there will no longer be non-identifier symbols in this list.  But every module defines an `eval` function, which currently has no docstring.

File: base/docs/Docs.jl [hunk 7]
```diff
@@ -675,4 +675,19 @@ function hasdoc(binding::Docs.Binding, sig::Type = Union{})
 end
 
 
+"""
+    undocumented_names(mod::Module; all=false)
+
+Return an array of undocumented symbols in `module` (that is, lacking docstrings).
+`all=false` returns only exported symbols; whereas `all=true` also includes
+non-exported symbols, following the behavior of [`names`](@ref).
```
Comment (stevengj) [inline base/docs/Docs.jl:683]: This, along with the `names` docstring, should mention that the names are sorted in ascending order (by `[isless](@ref)`, i.e. alphabetically).

File: base/docs/Docs.jl [hunk 8]
```diff
@@ -675,4 +675,19 @@ function hasdoc(binding::Docs.Binding, sig::Type = Union{})
 end
 
 
+"""
+    undocumented_names(mod::Module; all=false)
```
Comment (stevengj) [inline base/docs/Docs.jl:679]: needs to document `identifiers`.  Also should have a `::Bool` type qualifier on the keywords

Comment (jariji) [reply base/docs/Docs.jl:679]: I can't think of any use cases for `identifiers=false` so I just dropped it.

File: doc/src/manual/documentation.md (hunk 5 above)
Comment (jariji) [reply doc/src/manual/documentation.md:24]: Is there supposed to be a `Docs/index.md` somewhere? I don't see one.

Comment (stevengj) [reply doc/src/manual/documentation.md:24]: Probably we should add a section to https://github.com/JuliaLang/julia/blob/f1f680f5605ecc471678073391f280696640207e/doc/src/base/base.md, maybe after the `## Reflection` section, like:
````
## Documentation
(See also the [documentation](@ref man-documentation) chapter.)
```@docs
Base.@doc
Docs.doc
Docs.HTML
Docs.Text
Docs.hasdoc
Docs.undocumented_names
```
````

**Comment by jariji** [comment]
idk what that failure is.

**Review by stevengj — APPROVED** [review]

**Comment by jariji** [comment]
This is ready for merge.

**Comment by LilithHafner** [comment]
Sorry I'm a little late here; I'm not sure implementing this function in Base.Docs is a good idea. It seems like the kind of functionality that is used in development and testing but not in ordinary usage. Consequently, I think it belongs in Test or Aqua where it can be implemented in terms of Base's existing public API (thanks to @jariji's https://github.com/JuliaLang/julia/pull/52139) My opinion here is pretty weak, and I'd be open to possibly keeping it here, but I do think this should be discussed before we commit to adding this public API to Base.Docs

Regardless of the above, we should fix the docstring which currently misrepresents the behavior on undocumetned public but unexported symbols. (IIUC docstring: they are not returned by default; reality: they are returned by default)

**Comment by LilithHafner** [comment]
Clarification: IMO having this functionality is definitely a good idea, the question is _where_ to have it.

**Comment by stevengj** [comment]
If it goes anywhere else, it should be in Test (not Aqua) so that it can be used to test Base modules and stdlibs.  (That was my initial suggestion from #51174.)  I don't remember why we didn't discuss it here, sorry.

**Comment by stevengj** [comment]
I think the argument to put it in `Docs` was due to @Seelengrab in https://github.com/JuliaLang/julia/issues/51174#issuecomment-1841575684

**Comment by Seelengrab** [comment]
You're right, having `Docs` be a part of `Base` means the dependency is a non-issue.

Still, the thought was that `undocumented_names` is not related to testing functionality at all; there's a lot of things you could do with this (e.g. mark those names specially in Documenter, Pluto, some other tool) that aren't related to unit testing at all. Forcing that code to depend on `Test` in addition to `Docs` when the rest of the functionality of `Test` is not needed seems odd to me.

File: base/docs/Docs.jl [hunk 9]
```diff
@@ -675,4 +675,20 @@ function hasdoc(binding::Docs.Binding, sig::Type = Union{})
...
+Return an array of undocumented symbols in `module` (that is, lacking docstrings).
+`all=false` returns only exported symbols; whereas `all=true` also includes
+non-exported symbols, following the behavior of [`names`](@ref). Only valid identifiers
+are included. Names are returned in sorted order.
+
+See also: [`names`](@ref), [`Docs.hasdoc`](@ref), [`Base.isidentifier`](@ref).
+"""
+function undocumented_names(mod::Module; all::Bool=false)
+    filter!(names(mod; all)) do sym
+        !hasdoc(mod, sym) && Base.isidentifier(sym)
```
Comment (stevengj) [inline base/docs/Docs.jl:690]: I just realized that this filters out macros `@foo`, which we don't want.  Probably should be
```jl
str = string(sym)
!hasdoc(mod, sym) && (Base.isidentifier(sym) || (startswith(str, "@") && Base.isidentifier(str[2:end])))
```

**Comment by DilumAluthge** [comment]
Similar to the argument that I make in https://github.com/JuliaLang/julia/pull/52724, I think that we should rename `Docs.undocumented_names` to something a bit more specific and less ambiguous, to clarify that we are talking about names without docstrings.

I've opened https://github.com/JuliaLang/julia/pull/52726, which proposes renaming it to `Docs.names_without_docstring`, but I'm open to other names as well.

**Comment by LilithHafner** [comment]
From the dregs of triage (Lilith, Jeff, Jar) after most folks left:

Base.Docs is a fine place for this functionality to go. It's nontrivial to get right, so worth implementing, and it's needed by Base, so worth putting in Base/Docs/Test. It can be used even in the absence of testing, so shouldn't be in Test.

undocumented_names is a reasonable name. The only thing that "documented" could mean, from Base.Docs's perspective, is "has a docstring" because Base.Docs doesn't know about the manual or any Documenter.jl stuff.


```julia
_is_hidden_name(s::Symbol) = !isempty(string(x)) && string(x)[1] == '#
```

And use this semantic for `Docs.undocumented_names(mod; private=false)`:
For private=false, default: `filter(x -> !Base.Docs.hasdoc(mod, x) && !_is_hidden_name(x) && Base.ispublic(mod, x), names(mod, all=true))`
For private=true: `filter(x -> !Base.Docs.hasdoc(mod, x) && !_is_hidden_name(x), names(mod, all=true))`

Following the pattern of `names` which has keywords that default to false and setting them to true adds more names to be returned.

**Pull request merged**